modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBDecision.java

```diff
@@ -0,0 +1,310 @@
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+
+package org.geotools.mbstyle.expression;
+
+import org.geotools.mbstyle.parse.MBFormatException;
+import org.json.simple.JSONArray;
+import org.opengis.filter.expression.Expression;
+
+
+/**
+ * The expressions in this section can be used to add conditional logic to your styles. For example, the 'case'
+ * expression provides basic "if/then/else" logic, and 'match' allows you to map specific values of an input
+ * expression to different output expressions.
+ */
+public class MBDecision extends MBExpression {
+
+    // static String operators
+    private static final String NOT = "!";
+    private static final String NOT_EQUALS = "!=";
+    private static final String LESS_THAN = "<";
+    private static final String LESS_THAN_EQUALS = "<=";
+    private static final String EQUALS = "==";
+    private static final String GREATER_THAN = ">";
+    private static final String GREATER_THAN_EQUALS = ">=";
+    private static final String ALL = "all";
+    private static final String ANY = "any";
+    private static final String CASE = "case";
+    private static final String COALESCE = "coalesce";
+    private static final String MATCH = "match";
+
+    public MBDecision(JSONArray json) {
+        super(json);
+    }
+
+    /**
+     * Logical negation. Returns true if the input is false, and false if the input is true.
+     * Example: ["!", boolean]: boolean
+     * @return
+     */
+    private Expression decisionNot() {
+        // validate the arg list
+        if (json.size() != 2) {
+            throwUnexpectedArgumentCount(NOT, 1);
+        }
+        // second argument better be a Boolean, or another expression that results in a Boolean
+        Expression boolArg = parse.string(json, 1);
+        // return the opposite of the arg
+        return ff.function("not", boolArg);
+    }
+
+
+    /**
+     * Returns true if the input values are not equal, false otherwise.
+     * The inputs must be numbers, strings, or booleans, and both of the same type.
+     * Examples:["!=", number, number]: boolean
+     *          ["!=", string, string]: boolean
+     *          ["!=", boolean, boolean]: boolean
+     *          ["!=", null, null]: boolean
+     * @return
+     */
+    private Expression decisionNotEqual() {
+        // validate the arg list
+        if (json.size() != 3) {
+            throwUnexpectedArgumentCount(NOT_EQUALS, 2);
+        }
+        // get the comparables
+        Expression comparable1 = parse.string(json, 1);
+        Expression comparable2 = parse.string(json, 2);
+        return ff.function("mbNotEqualTo", comparable1, comparable2);
```

Comment:
why make a new "mbNotEqualTo" function when you can use "notEqualTo"?

See geotools function list entry here: http://docs.geotools.org/latest/userguide/library/main/function_list.html#notequalto-a-b-notequalto

Reply:
Reading your description there, the handling of null != null sounds indeed different. The fallback comparison based on strings may be a bug in the implementation of "notEqualTo".

Reply:
@jodygarnett The GeoTools implementations for "notEqualTo" and "equalTo" (and the StaticGeometry methods they call) haven't changed since early 2011. While it may be a bug/oversight, I would assume things relying on those functions expect them to behave as they are, buggy or not.